modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -3230,13 +3230,19 @@ def to_pandas(self):
         if df.empty:
             df = pandas.DataFrame(columns=self.columns, index=self.index)
         else:
-            for axis in [0, 1]:
-                ErrorMessage.catch_bugs_and_request_email(
-                    not df.axes[axis].equals(self.axes[axis]),
-                    f"Internal and external indices on axis {axis} do not match.",
-                )
-            df.index = self.index
-            df.columns = self.columns
+            for axis, public_index in enumerate(
+                [self._index_cache, self._columns_cache]
+            ):
+                # no need to check external and internal axes since in that case
+                # external axes will be computed from internal partitions
+                if public_index is not None:
+                    ErrorMessage.catch_bugs_and_request_email(
+                        not df.axes[axis].equals(public_index),
+                        f"Internal and external indices on axis {axis} do not match.",
+                    )
+                    df.set_axis(
```

Comment:
Why do we need to call `set_axis` if we know the axes are equal?

Reply:
this is done in order to assign some potentially missing metadata, the ones that were set to the external index but were never propagated into the internal ones:
```python
>>> pd.Index([1], name="hey").equals(pd.Index([1]))
True
```

Reply:
I see, thanks! Let's add a comment why we do this because it might be confusing for some developers.

Reply:
@YarShev added